docs/developer-guide/ksqldb-clients/java-client.md

```diff
@@ -0,0 +1,416 @@
+  // Block until a new row is available
+  final Row row = streamedQueryResult.poll();
+  if (row != null) {
+    System.out.println("Received a row!");
+    System.out.println("Row: " + row.values());
+  } else {
+    System.out.println("Query has ended.");
+  }
+}
+```
+
+Receive query results in a single batch (executeQuery())<a name="execute-query"></a>
+------------------------------------------------------------------------------------
+
+The `executeQuery()` method allows client apps to receive query results as a single batch,
+returned once the query has completed.
+
+```java
+public interface Client {
+
+  /**
+   * Executes a query (push or pull) and returns all result rows in a single batch, once the query
+   * has completed.
+   *
+   * @param sql statement of query to execute
+   * @return query result
+   */
+  BatchedQueryResult executeQuery(String sql);
+  
+  ...
+  
+}
+```
+
+This method is suitable for both pull queries as well as terminating push queries (e.g., those with a `LIMIT` clause).
+For non-temrinating push queries, use [the `streamQuery()` method](./stream-query.md) instead. 
+
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
+
+By default, push queries return results starting from the beginning of the stream or table.
+To instead start from the end and only receive newly arriving rows, set the property `auto.offset.reset` to `latest`.
+
+### Example Usage ###
+
+```java
+final String pullQuery = "SELECT * FROM MY_MATERIALIZED_TABLE WHERE KEY_FIELD='some_key';";
+final BatchedQueryResult batchedQueryResult = client.executeQuery(pullQuery);
+
+final List<Row> resultRows;
+try {
+  resultRows = batchedQueryResult.get();
+} catch (Exception e) {
+  System.out.println("Request failed: " + e);
+  return;
+}
+
+System.out.println("Received results. Num rows: " + resultRows.size());
+for (final Row row : resultRows) {
+  System.out.println("Row: " + row.values());
+}
+```
+
+Terminate a push query (terminatePushQuery())<a name="terminate-push-query"></a>
+--------------------------------------------------------------------------------
+
+The `terminatePushQuery()` method allows client apps to terminate push queries.
+
+```java
+public interface Client {
+
+  /**
+   * Terminates a push query with the specified query ID.
+   *
+   * <p>If a non-200 response is received from the server, the {@code CompletableFuture} will be
+   * failed.
+   *
+   * @param queryId ID of the query to terminate
+   * @return a future that completes once the server response is received
+   */
+  CompletableFuture<Void> terminatePushQuery(String queryId);
+  
+  ...
+  
+}
+```
+
+The query ID is obtained from the query result response object when push queries are issued via the client,
+via either [`streamQuery()`](./stream-query.md) or [`executeQuery()`](./execute-query.md).
+
+### Example Usage ###
+
+Here's an example of terminating a push query issued via `streamQuery()`:
```

Comment:
```suggestion
Here's an example of terminating a push query issued by using the `streamQuery()` method:
```

Reply:
Out of curiosity, why the preference for "a push query issued by using the `streamQuery()` method" rather than "a push query issued using the `streamQuery()` method"? I've currently updated to the latter but can easily switch to your original suggestion instead. (Can't quite pin it but the extra "by" reads weirdly to me, haha)

Reply:
I think strictly speaking (grammatically) Jim is right... (?)

I.e. it seems more natural for me to say "a push query issued _by_ using the.." or "a push query issued _while_ using the..." It seems to me it needs some word in there.

Having said that.... it probably doesn't matter a great deal ;)

https://forum.wordreference.com/threads/relative-pronouns-omitting-relative-pronoun-not-relative-adverb.282370/